Add WeChat Receiver type to AlertmanagerConfig CRD

zwpaper opened the pull request:
fix https://github.com/prometheus-operator/prometheus-operator/issues/3537

btw, I updated the user guide doc to use AlertmanagerConfig CRD

grdryn commented on Documentation/user-guides/alerting.md:
```diff
@@ -30,9 +32,72 @@ spec:
 
+## AlertmanagerConfig Resource
+
+The following example configuration creates an AlertmanagerConfig resource that sends notifications against a non-existent `wechat` receiver, allowing the Alertmanager to start up, without issuing any notifications.
```
```suggestion
The following example configuration creates an AlertmanagerConfig resource that sends notifications to a non-existent `wechat` receiver:
```

(removing the part of the sentence that suggests that something like this would be necessary to allow Alertmanager to start)

grdryn commented on Documentation/user-guides/alerting.md:
````diff
@@ -30,9 +32,72 @@ spec:
+$ kubectl create -f alertmanager-config.yaml
+```
+
+Alertmanager require a `alertmanagerConfigSelector` for finding those AlertmanagerConfig resources. In the previous example, the label `alertmanagerConfig: example` is added, so Alertmanager instance should be updated, adding the `alertmanagerConfigSelector`:
````
```suggestion
The `alertmanagerConfigSelector` field in the Alertmanager resource Spec needs to be specified so that the operator can select such AlertmanagerConfig resources. In the previous example, the label `alertmanagerConfig: example` is added, so the Alertmanager instance should be updated, adding the `alertmanagerConfigSelector`:
```

grdryn commented on Documentation/user-guides/alerting.md:
```diff
@@ -30,9 +32,72 @@ spec:
+
+## Manually Managed Secret
+
+The following example configuration sends notifications against a non-existent `webhook`, allowing the Alertmanager to start up, without issuing any notifications.
```
```suggestion
The following example configuration sends notifications to a non-existent `webhook`:
```

grdryn commented on Documentation/user-guides/alerting.md:
```diff
@@ -79,7 +144,9 @@ templates:
-Once created this Secret is mounted by Alertmanager Pods created through the Alertmanager object.
+## Expose Alertmanager
+
+Once the `AlertmanagerConfig` or the configuration Secret is mounted by Alertmanager Pods created through the Alertmanager object.
```
```suggestion
Once the operator merges the optional manually specified Secret with any selected `AlertmanagerConfig` resources, a new configuration Secret is created with the name `alertmanager-<Alertmanager name>-generated`, and is mounted into Alertmanager Pods created through the Alertmanager object.
```

grdryn commented on pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:
```diff
@@ -168,6 +169,18 @@ type PagerDutyConfigDetail struct {
 	Value string `json:"value"`
 }
 
+type WeChatConfig struct {
```
There are a couple of extra properties in the upstream type that are not reflected here:

* [HTTPConfig](https://github.com/prometheus/alertmanager/blob/master/config/notifiers.go#L397)
* [MessageType](https://github.com/prometheus/alertmanager/blob/master/config/notifiers.go#L407)

Do you plan to add those?

zwpaper replied:
my bad, I only checked the alertmanager doc website, not finding those two configs, added!

grdryn reviewed (commented):
Thanks for updating the doc, it seems I completely missed that when adding the initial AlertmanagerConfig stuff! I've suggested a couple of changes to your proposed changes here.

zwpaper commented on Documentation/user-guides/alerting.md:
```diff
@@ -18,6 +18,8 @@ The Alertmanager may be used to:
 
 Prometheus' configuration also includes "rule files", which contain the [alerting rules][alerting-rules]. When an alerting rule triggers it fires that alert against *all* Alertmanager instances, on *every* rule evaluation interval. The Alertmanager instances communicate to each other which notifications have already been sent out. For more information on this system design, see the [High Availabi ...
 
+The Prometheus Operator also introduces an AlertmanagerConfig resource, which allows users to declaratively describe Alertmanager configurations. It is recommended to use the AlertmanagerConfig resource than manage the configurations manually.
```
how about changing this to

```
The AlertmanagerConfig resource is currently v1alpha1, testing and feedback are welcome.
```

grdryn replied:
Sure, that sounds better to me :+1: Will leave it up to the project maintainers to review and have the final say :slightly_smiling_face:

zwpaper commented:
hi @grdryn , I have changed the doc and added the HTTPConfig and MessageType, thanks for the review.

after I click the re-request review from you, the reviewer `prometheus-operator-reviewer` is gone, is this as expected?

grdryn commented:
@zwpaper thanks for addressing my comments, I don't have any more. Maybe one of the 

> after I click the re-request review from you, the reviewer prometheus-operator-reviewer is gone, is this as expected?

Hmm, I'm not sure, I don't know anything about that really. Maybe @simonpasquier or @lilic might know how to get that back if it's important? :slightly_smiling_face:

zwpaper commented:
@grdryn It's not that important if the project maintainers can notice the PR😃

lilic reviewed (approved):
Very nice, thanks so much!!

lgtm
